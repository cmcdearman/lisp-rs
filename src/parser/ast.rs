use num_rational::Rational64;

use crate::{intern::InternedString, list::List};

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Symbol(InternedString),
    Lit(Lit),
    List(List<Self>),
    Lambda {
        param: Vec<Self>,
        body: Box<Self>,
    },
    Apply {
        func: Box<Self>,
        arg: Box<Self>,
    },
    Let {
        name: InternedString,
        value: Box<Self>,
        body: Box<Self>,
    },
    If {
        cond: Box<Self>,
        then: Box<Self>,
        else_: Box<Self>,
    },
    Unit,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Lit {
    Int(i64),
    Real(f64),
    Rational(Rational64),
    Bool(bool),
    Char(char),
    String(InternedString),
}
