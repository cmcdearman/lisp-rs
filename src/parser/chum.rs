use crate::T;

use super::{sexpr::{Sexpr, Number}, token::Token};
use chumsky::{prelude::Simple, recursive::recursive, Parser, select, primitive::just};

pub fn parser() -> impl Parser<Token, Sexpr, Error = Simple<Token>> {
    recursive(|sexpr| {
        let number = select! {
            T![int] => Number::Fixnum(),
            T![float] => Sexpr::Number(Number::Float(0.0)),
        };

        let lit = select! {
            T![int] => Sexpr::Number(Number::Int(0)),
            T![float] => Sexpr::Number(Number::Float(0.0)),
            T![string] => Sexpr::String("".to_string()),
            T![char] => Sexpr::Char('a'),
            T![bool] => Sexpr::Bool(true),
        };
        
        let atom = select! {
            T![int] => 
        }

        atom
    })
}
