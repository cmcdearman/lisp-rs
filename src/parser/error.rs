use std::fmt::Display;

use crate::intern::InternedString;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Error(pub InternedString);

impl Error {
    pub fn new(msg: &str) -> Self {
        Self(msg.into())
    }
}

impl Display for Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        InternedString::from(self.0.key).fmt(f)
    }
}

impl From<String> for Error {
    fn from(s: String) -> Self {
        Self(InternedString::from(&*s))
    }
}

pub type ParseResult<T> = std::result::Result<T, Error>;
