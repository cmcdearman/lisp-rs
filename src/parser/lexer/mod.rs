use logos::Logos;

use self::token::{TokenKind, Token};

pub mod token;

pub struct Lexer<'src> {
    logos: logos::SpannedIter<'src, TokenKind>,
    skip_comments: bool,
    eof: bool,
}

impl<'src> Lexer<'src> {
    pub fn new(src: &'src str, skip_comments: bool) -> Self {
        Self {
            logos: TokenKind::lexer(src).spanned(),
            skip_comments,
            eof: false,
        }
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        match self.logos.next() {
            Some((t, s)) => {
                if t == TokenKind::Comment && self.skip_comments {
                    return self.next();
                } else {
                    return Some(Token {
                        kind: t,
                        span: s.into(),
                    });
                }
            }
            None if self.eof => None,
            None => {
                self.eof = true;
                Some(Token {
                    kind: TokenKind::Eof,
                    span: (0..0).into(),
                })
            }
        }
    }
}

// mod tests {
//     #[test]
//     fn test_lex() {
//         let input = "(+ 3 4) ; test comment";
//     }
// }