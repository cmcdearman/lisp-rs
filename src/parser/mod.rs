use std::{cell::RefCell, iter::Peekable, rc::Rc, vec::IntoIter};

use either::Either;

use crate::T;

use sexpr::{Atom, Lit, Sexpr};

use self::{
    error::{ParserError, Result},
    lexer::Lexer,
    token::{Token, TokenKind},
};

pub mod error;
pub mod lexer;
pub mod sexpr;
pub mod token;

pub struct Parser<'src> {
    src: &'src str,
    tokens: Peekable<Either<Lexer<'src>, IntoIter<Token>>>,
}

impl<'src> Parser<'src> {
    pub fn new(src: &'src str, lazy: bool) -> Self {
        Self {
            src,
            tokens: if lazy {
                Either::Left(Lexer::new(src, true)).peekable()
            } else {
                Either::Right(Lexer::new(src, true).collect::<Vec<Token>>().into_iter()).peekable()
            },
        }
    }

    fn text(&self, token: Token) -> &'src str {
        token.lit(&self.src)
    }

    fn peek(&mut self) -> TokenKind {
        self.tokens
            .peek()
            .map(|token| token.kind)
            .unwrap_or(T![EOF])
    }

    fn at(&mut self, kind: TokenKind) -> bool {
        self.peek() == kind
    }

    fn next(&mut self) -> Option<Token> {
        self.tokens.next()
    }

    fn consume(&mut self, expected: TokenKind) {
        let token = self.next().expect(&format!(
            "Expected to consume `{}`, but there was no next token",
            expected
        ));
        assert_eq!(
            token.kind, expected,
            "Expected to consume `{}`, but found `{}`",
            expected, token.kind
        );
    }

    pub fn sexpr(&mut self) -> Result<Sexpr> {
        match self.peek() {
            T!['('] => {
                self.consume(T!['(']);
                self.list()
            }
            _ => self.atom(),
        }
    }

    fn list(&mut self) -> Result<Sexpr> {
        let mut new_list = List { head: None };
        let mut tail: Option<Rc<RefCell<Cons>>> = None;

        while !self.at(T![')']) {
            let new_cons = Rc::new(RefCell::new(Cons {
                car: self.sexpr()?,
                cdr: None,
            }));
            if new_list.head.is_none() {
                new_list.head = Some(new_cons.clone());
            } else if let Some(tail_cons) = tail {
                tail_cons.as_ref().borrow_mut().cdr = Some(new_cons.clone());
            }

            tail = Some(new_cons);
        }

        Ok(Sexpr::List(new_list))
    }

    fn atom(&mut self) -> Result<Sexpr> {
        match self.peek() {
            lit @ T![int] | lit @ T![float] | lit @ T![str] | lit @ T![bool] => {
                let lit_text = {
                    let lit_tok = self.next().expect("expected token but found None");
                    self.text(lit_tok)
                };
                let lit = match lit {
                    T![int] => Lit::Number(Number::Integer(
                        lit_text.parse::<Integer>().map_err(|e| ParserError(e.0))?,
                    )),
                    T![float] => Lit::Number(Number::Float(
                        lit_text.parse::<Float>().map_err(|e| ParserError(e.0))?,
                    )),
                    T![str] => Lit::Str(lit_text[1..(lit_text.len() - 1)].to_string()),
                    T![bool] => Lit::Bool(lit_text.parse().expect("invalid bool literal")),
                    _ => unreachable!(),
                };
                Ok(Sexpr::Atom(Atom::Lit(lit)))
            }
            TokenKind::Ident => {
                let ident = self.next().ok_or(ParserError::new("t".to_string()))?;
                Ok(Sexpr::Atom(Atom::Sym(Symbol::from(self.text(ident)))))
            }
            kind => {
                panic!("Unknown start of atom: `{}`", kind);
            }
        }
    }
}
