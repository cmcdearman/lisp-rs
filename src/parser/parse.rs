use std::{cell::RefCell, iter::Peekable, rc::Rc};

use crate::{
    object::{
        cons::Cons,
        list::List,
        number::{
            integer::{fixnum::FixNum, Integer},
            Number,
        },
        symbol::Symbol,
        Atom, Lit, Object,
    },
    token::{TokenKind, TokenStream},
};

pub mod error;
pub mod lexer;
pub mod token;

pub struct Parser<'src> {
    src: &'src str,
    tokens: Peekable<Either<Lexer<'src>, IntoIter<Token>>>,
}

impl<'src> Parser<'src> {
    pub fn new(src: &'src str, lazy: bool) -> Self {
        Self {
            src,
            tokens: if lazy {
                Either::Left(Lexer::new(src, true)).peekable()
            } else {
                Either::Right(Lexer::new(src, true).collect::<Vec<Token>>().into_iter()).peekable()
            },
        }
    }

    fn text(&self, token: Token) -> &'src str {
        token.lit(&self.src)
    }

    fn peek(&mut self) -> TokenKind {
        self.tokens
            .peek()
            .map(|token| token.kind)
            .unwrap_or(T![EOF])
    }

    fn at(&mut self, kind: TokenKind) -> bool {
        self.peek() == kind
    }

    fn next(&mut self) -> Option<Token> {
        self.tokens.next()
    }

    fn consume(&mut self, expected: TokenKind) {
        let token = self.next().expect(&format!(
            "Expected to consume `{}`, but there was no next token",
            expected
        ));
        assert_eq!(
            token.kind, expected,
            "Expected to consume `{}`, but found `{}`",
            expected, token.kind
        );
    }

    pub fn sexpr(tokens: &mut Peekable<TokenStream>) -> Result<Object, String> {
        match tokens.peek().unwrap().kind {
            TokenKind::LParen => {
                tokens.next();
                list(tokens)
            }
            _ => atom(tokens),
        }
    }

    fn list(tokens: &mut Peekable<TokenStream>) -> Result<Object, String> {
        let mut new_list = List { head: None };
        let mut tail: Option<Rc<RefCell<Cons>>> = None;

        while tokens.peek().unwrap().kind != TokenKind::RParen {
            let new_cons = Rc::new(RefCell::new(Cons {
                car: parse(tokens)?,
                cdr: None,
            }));
            if new_list.head.is_none() {
                new_list.head = Some(new_cons.clone());
            } else if let Some(tail_cons) = tail {
                tail_cons.as_ref().borrow_mut().cdr = Some(new_cons.clone());
            }

            tail = Some(new_cons);
        }

        Ok(Object::List(new_list))
    }

    fn atom(tokens: &mut Peekable<TokenStream>) -> Result<Object, String> {
        match tokens.peek().unwrap().kind {
            lit @ TokenKind::Int
            | lit @ TokenKind::Float
            | lit @ TokenKind::String
            | lit @ TokenKind::Bool => {
                let lit_text = tokens.next().unwrap().lit;
                let lit = match lit {
                    TokenKind::Int => Lit::Num(Number::Integer(Integer::FixNum(FixNum(
                        lit_text
                            .parse()
                            .expect(&format!("invalid floating point literal: `{}`", lit_text)),
                    )))),
                    TokenKind::Float => Lit::Num(Number::Float(
                        lit_text
                            .parse()
                            .expect(&format!("invalid floating point literal: `{}`", lit_text)),
                    )),
                    TokenKind::String => Lit::Str(lit_text[1..(lit_text.len() - 1)].to_string()),
                    TokenKind::Bool => Lit::Bool(lit_text.parse().unwrap()),
                    _ => unreachable!(),
                };
                Ok(Object::Atom(Atom::Lit(lit)))
            }
            TokenKind::Ident => Ok(Object::Atom(Atom::Sym(Symbol::from(
                &*tokens.next().ok_or("end of tokens".to_string())?.lit,
            )))),
            kind => {
                panic!("Unknown start of atom: `{}`", kind);
            }
        }
    }
}
