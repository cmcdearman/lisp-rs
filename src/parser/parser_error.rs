use std::fmt::Display;

use super::token::Span;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum ParserErrorKind {
    ParseIntegerError,
    ParseFloatError,
    ParseRationalError,
    ParseStringError,
    UnexpectedEofError,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct ParserError {
    kind: ParserErrorKind,
    span: Span,
}

impl ParserError {
    pub fn new(kind: ParserErrorKind, span: Span) -> Self {
        Self { kind, span }
    }
}

impl Display for ParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}", self)
    }
}

pub type Result<T> = std::result::Result<T, ParserError>;
