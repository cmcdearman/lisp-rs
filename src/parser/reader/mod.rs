use crate::{intern::InternedString, T};
use logos::{Lexer, Logos};
use num_rational::Rational64;
use std::fmt::Debug;

use self::{
    error::{Error, ReadResult},
    sexpr::{Atom, Cons, ConsList, Lit, Sexpr},
    token::{Span, Token, TokenKind},
};

pub mod error;
pub mod sexpr;
mod tests;
pub mod token;

pub struct Reader<'src> {
    src: &'src str,
    logos: Lexer<'src, TokenKind>,
    peek: Option<Token>,
    errors: Vec<Error>,
}

impl<'src> Reader<'src> {
    /// Creates a new [`Parser`].
    pub fn new(src: &'src str) -> Self {
        Self {
            src,
            logos: TokenKind::lexer(src),
            peek: None,
            errors: vec![],
        }
    }

    /// Returns the source code of the token.
    fn text(&self, token: Token) -> &'src str {
        token.lit(&self.src)
    }

    /// Returns the peek token in the stream.
    fn next(&mut self) -> Token {
        if let Some(t) = self.peek.take() {
            t
        } else {
            self.fetch_token()
        }
    }

    /// Returns the next token in the stream without consuming it.
    fn peek(&mut self) -> Token {
        if let Some(t) = self.peek {
            t
        } else {
            let t = self.fetch_token();
            self.peek = Some(t);
            t
        }
    }


    /// Returns true if the next token is of the given kind.
    fn at(&mut self, kind: TokenKind) -> bool {
        self.peek().kind == kind
    }

    /// Consumes the next token if it is of the given kind.
    fn eat(&mut self, expected: TokenKind) -> bool {
        if self.at(expected) {
            self.next();
            true
        } else {
            false
        }
    }

    /// Parses the source code into a [`Sexpr`].
    pub fn sexpr(&mut self) -> ReadResult<Sexpr> {
        match self.peek().kind {
            T!['('] => self.list(),
            _ => self.atom(),
        }
    }

    fn list(&mut self) -> ReadResult<Sexpr> {
        if !self.eat(T!['(']) {
            return Err(Error::new("Expected `(`"));
        }
        let mut sexprs = vec![];
        while !self.at(T![')']) {
            sexprs.push(self.sexpr()?);
        }
        if !self.eat(T![')']) {
            return Err(Error::new("Expected `)`"));
        }
        Ok(Sexpr::List(ConsList::new(
            sexprs
                .into_iter()
                .rev()
                .fold(None, |cdr, car| Some(Box::new(Cons { car, cdr }))),
        )))
    }

    fn atom(&mut self) -> ReadResult<Sexpr> {
        match self.peek().kind {
            T![int] | T![real] | T![ratio] | T![char] | T![str] | T![bool] => {
                Ok(Sexpr::Atom(Atom::Lit(self.lit()?)))
            }
            T![ident] => Ok(Sexpr::Atom(Atom::Symbol(self.symbol()?))),
            _ => Err(Error::from(format!(
                "Unexpected token in atom `{}`",
                self.peek().kind
            ))),
        }
    }

    fn lit(&mut self) -> ReadResult<Lit> {
        match self.peek().kind {
            T![int] => Ok(Lit::Int(self.int()?)),
            T![real] => Ok(Lit::Real(self.real()?)),
            T![ratio] => Ok(Lit::Rational(self.rational()?)),
            T![char] => Ok(Lit::Char(self.char()?)),
            T![str] => Ok(Lit::String(self.string()?)),
            T![bool] => Ok(Lit::Bool(self.bool()?)),
            _ => Err(Error::from(format!(
                "Unexpected token in literal `{}`",
                self.peek().kind
            ))),
        }
    }

    fn int(&mut self) -> ReadResult<i64> {
        let token = self.next();
        let text = self.text(token);
        let num = text
            .parse()
            .map_err(|_| Error::new("Failed to parse integer"))?;
        Ok(num)
    }

    fn real(&mut self) -> ReadResult<f64> {
        let token = self.next();
        let text = self.text(token);
        let num = text
            .parse()
            .map_err(|_| Error::new("Failed to parse float"))?;
        Ok(num)
    }

    fn rational(&mut self) -> ReadResult<Rational64> {
        let token = self.next();
        let text = self.text(token);
        let num = text
            .parse()
            .map_err(|_| Error::new("Failed to parse rational"))?;
        Ok(num)
    }

    fn string(&mut self) -> ReadResult<InternedString> {
        let token = self.next();
        let text = self.text(token);
        Ok(InternedString::from(&text[1..(text.len() - 1)]))
    }

    fn char(&mut self) -> ReadResult<char> {
        let token = self.next();
        let text = self.text(token);
        Ok(text
            .chars()
            .nth(1)
            .ok_or(Error::new("Failed to parse char"))?)
    }

    fn bool(&mut self) -> ReadResult<bool> {
        let token = self.next();
        let text = self.text(token);
        Ok(text == "#t")
    }

    fn symbol(&mut self) -> ReadResult<InternedString> {
        let token = self.next();
        let text = self.text(token);
        Ok(InternedString::from(text))
    }
}
