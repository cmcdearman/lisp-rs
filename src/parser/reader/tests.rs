use crate::parser::Reader;

#[test]
fn test_read_int() {
    let mut reader = Reader::new("12");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_float() {
    let mut reader = Reader::new("12.2");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_rational() {
    let mut reader = Reader::new("12/2");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_char() {
    let mut reader = Reader::new("\'a\'");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_string() {
    let mut reader = Reader::new("\"Hello, world!\"");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_bool() {
    let mut reader = Reader::new("true");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_symbol() {
    let mut reader = Reader::new("foo");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_add_list() {
    let mut reader = Reader::new("(+ 1 2)");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}

#[test]
fn test_read_nested_list() {
    let mut reader = Reader::new("(+ (% 5 3) 1 2)");
    let sexpr = reader.sexpr().expect("Failed to read sexpr");
    insta::assert_debug_snapshot!(sexpr);
}
