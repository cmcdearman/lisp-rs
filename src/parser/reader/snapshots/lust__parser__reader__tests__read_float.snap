---
source: src/parser/reader/tests.rs
expression: sexpr
---
12.2
